File: dep/guifast/frontend/src/shared/action/libflo_initialized.ts
import * as Guifast from "guifast_shared";
import * as GuifastShared from "guifast/shared";

export namespace LibfloInitialized {
    const str = "libflo_initialized";
    export const id = Guifast.makeStrId(Guifast.String.module, str);

    export interface Action extends Guifast.Action {
        readonly action_mapper: GuifastShared.ActionMapperSerde,
        readonly config: GuifastShared.GuifastConfigSerde,
        readonly module_mapper: GuifastShared.ModuleMapperSerde,
        readonly modules: GuifastShared.ModulesSerde
    }

    export const make = (
        actionMapper: GuifastShared.ActionMapperSerde,
        config: GuifastShared.GuifastConfigSerde,
        moduleMapper: GuifastShared.ModuleMapperSerde,
        modules: GuifastShared.ModulesSerde
    ): Action => {
        return {
            type: id,
            action_mapper: actionMapper,
            config: config,
            module_mapper: moduleMapper,
            modules: modules
        };
    }; 
}

File: dep/guifast_shared/backend/src/action/libflo_initialized.rs
use libflo_action_std::{ Action, ActionMapperSerde, action_str_id, NumberOrString };
use libflo_std::ModuleMapperSerde;
use serialization::{ GuifastConfigSerde, ModulesSerde };
use string;

#[derive(Clone, Debug, Serialize)]
pub struct LibfloInitialized<'a, 'b, 'c, 'd, 'e> {
    #[serde(rename = "type")]
    action_type: NumberOrString,
    action_mapper: ActionMapperSerde<'a, 'b>,
    config: &'c GuifastConfigSerde,
    module_mapper: ModuleMapperSerde<'d, 'e>,
    modules: ModulesSerde,
}

impl <'a, 'b, 'c, 'd, 'e> LibfloInitialized<'a, 'b, 'c, 'd, 'e> {
    pub fn new(action_mapper: ActionMapperSerde<'a, 'b>, config: &'c GuifastConfigSerde, module_mapper: ModuleMapperSerde<'d, 'e>, modules: ModulesSerde) -> Self {
        LibfloInitialized {
            action_type: action_str_id(string::module(), string::libflo_initialized_action()),
            action_mapper: action_mapper,
            config: config,
            module_mapper: module_mapper,
            modules: modules,
        }
    }
}

impl <'a, 'b, 'c, 'd, 'e> Action for LibfloInitialized<'a, 'b, 'c, 'd, 'e> {
    fn get_type(&self) -> &NumberOrString {
        &self.action_type
    }
}
